### PR TITLE
KAFKA-17600: Add nextOffsets to the ConsumerRecords

### PR DESCRIPTION
**Pull request by aliehsaeedii**
This PR implements KIP-1094.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 1]
```diff
@@ -53,6 +63,13 @@ public List<ConsumerRecord<K, V>> records(TopicPartition partition) {
             return Collections.unmodifiableList(recs);
     }
 
+    /**
+     * Get just the next offsets that the consumer will consume
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: I think this comment needs to be more precise. Which topic partitions are included, all the partitions that are assigned to the consumer, all partitions that I fetched from in the last batch? I suppose it's the latter

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 2]
```diff
@@ -53,6 +63,13 @@ public List<ConsumerRecord<K, V>> records(TopicPartition partition) {
...
+     */
+    public Map<TopicPartition, OffsetAndMetadata> nextOffsets() {
+        return nextOffsets;
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java:68]: Other methods in this class return an unmodifiable collection, should we do the same?

File: clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java [hunk 3]
```diff
@@ -64,41 +64,11 @@ public void testSimpleMock() {
...
         final TopicPartition tp = new TopicPartition("test", 0);
         assertEquals(2L, consumer.position(tp));
+        assertEquals(new OffsetAndMetadata(2, Optional.empty(), ""), recs.nextOffsets().get(tp));
         consumer.commitSync();
         assertEquals(2L, consumer.committed(Collections.singleton(tp)).get(tp).offset());
     }
 
-    @SuppressWarnings("deprecation")
-    @Test
-    public void testSimpleMockDeprecated() {
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java]: why did you remove this test?

File: clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java [hunk 4]
```diff
@@ -125,6 +95,7 @@ public void shouldNotClearRecordsForPausedPartitions() {
         consumer.resume(testPartitionList);
         ConsumerRecords<String, String> recordsSecondPoll = consumer.poll(Duration.ofMillis(1));
         assertEquals(1, recordsSecondPoll.count());
+        assertEquals(new OffsetAndMetadata(1, Optional.empty(), ""), recordsSecondPoll.nextOffsets().get(new TopicPartition("test", 0)));
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java:100]: We may want to compare all elements of `recordsSecondPoll.nextOffsets`, or at least check the size of the collection.

File: clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java [hunk 5]
```diff
@@ -64,41 +64,11 @@ public void testSimpleMock() {
         assertFalse(iter.hasNext());
         final TopicPartition tp = new TopicPartition("test", 0);
         assertEquals(2L, consumer.position(tp));
+        assertEquals(new OffsetAndMetadata(2, Optional.empty(), ""), recs.nextOffsets().get(tp));
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java:68]: We may want to compare all elements of `recs.nextOffsets`, or at least check the size of the collection.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java [hunk 6]
```diff
@@ -480,7 +481,7 @@ public ConsumerRecords<byte[], byte[]> consume(int n, long maxDuration, Map<Stri
                     consumedRecords += r.size();
                 }
                 if (consumedRecords >= n) {
-                    return new ConsumerRecords<>(records);
+                    return new ConsumerRecords<>(records, rec.nextOffsets());
```
Comment (lucasbru) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: Not sure if this is correct. Aren't you dropping all the `nextOffsets` for the previous polls? Like you are aggregating `rec.records` in `records`, you should aggregate `rec.nextOffsets` somewhere.

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java [hunk 7]
```diff
@@ -178,6 +179,7 @@ private ConsumerRecords<Integer, String> buildTopicTestRecords(int recordSize,
                                                                    int emptyPartitionIndex,
                                                                    Collection<String> topics) {
         Map<TopicPartition, List<ConsumerRecord<Integer, String>>> partitionToRecords = new LinkedHashMap<>();
+        Map<TopicPartition, OffsetAndMetadata> nextOffsets = new HashMap<>();
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java:188]: In `ConsumerRecordsTest`, you insert data but never check anything on it. Don't you want to verify something?

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 8]
```diff
@@ -21,23 +21,33 @@
 
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
+
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: extra new line

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockShareConsumer.java [hunk 9]
```diff
@@ -91,7 +91,7 @@ public synchronized ConsumerRecords<K, V> poll(Duration timeout) {
         }
 
         records.clear();
-        return new ConsumerRecords<>(results);
+        return new ConsumerRecords<>(results, Collections.emptyMap());
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockShareConsumer.java]: Could you check the result of `nextOffsets` in `MockShareConsumerTest` ? Generally, try to accompany each change with a change in the unit tests, if possible.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java [hunk 10]
```diff
@@ -95,6 +102,13 @@ public int numRecords() {
         return numRecords;
     }
 
+    /**
+     * @return the next offsets and metadata (last epochs is included)
+     */
+    public Map<TopicPartition, OffsetAndMetadata> nextOffsetAndMetadata() {
+        return nextOffsetAndMetadata;
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java]: `records` returns an unmodifiableMap. Should we do the same here?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java [hunk 11]
```diff
@@ -195,7 +196,7 @@ private Fetch<K, V> fetchRecords(final CompletedFetch nextInLineFetch, int maxRe
                     metricsManager.recordPartitionLead(tp, lead);
                 }
 
-                return Fetch.forPartition(tp, partRecords, positionAdvanced);
+                return Fetch.forPartition(tp, partRecords, positionAdvanced, new OffsetAndMetadata(nextInLineFetch.nextFetchOffset(), nextInLineFetch.lastEpoch(), ""));
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:199]: We _definitely_ need to update / add tests for this change in `FetchCollectorTest`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java [hunk 12]
```diff
@@ -583,7 +583,7 @@ public synchronized ConsumerRecords<K, V> poll(final Duration timeout) {
                 final ShareFetch<K, V> fetch = pollForFetches(timer);
                 if (!fetch.isEmpty()) {
                     currentFetch = fetch;
-                    return new ConsumerRecords<>(fetch.records());
+                    return new ConsumerRecords<>(fetch.records(), Collections.emptyMap());
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java]: Trivial, but please test this in `ShareConsumerImplTest`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerInterceptorsTest.java [hunk 13]
```diff
@@ -126,9 +131,15 @@ public void testOnConsumeChain() {
...
+        nextOffsets.put(filterTopicPart1, new OffsetAndMetadata(1, Optional.empty(), ""));
+
         records.put(filterTopicPart2, list3);
-        ConsumerRecords<Integer, Integer> consumerRecords = new ConsumerRecords<>(records);
+        nextOffsets.put(filterTopicPart2, new OffsetAndMetadata(1, Optional.empty(), ""));
+
+        ConsumerRecords<Integer, Integer> consumerRecords = new ConsumerRecords<>(records, nextOffsets);
         ConsumerRecords<Integer, Integer> interceptedRecords = interceptors.onConsume(consumerRecords);
         assertEquals(1, interceptedRecords.count());
         assertTrue(interceptedRecords.partitions().contains(tp));
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerInterceptorsTest.java:145]: I think we need to assert here something about the result of nextOffsets right? That those topic partitions are filtered out?

File: clients/src/test/java/org/apache/kafka/test/MockConsumerInterceptor.java [hunk 14]
```diff
@@ -73,19 +73,24 @@ public ConsumerRecords<String, String> onConsume(ConsumerRecords<String, String>
         CLUSTER_ID_BEFORE_ON_CONSUME.compareAndSet(NO_CLUSTER_ID, CLUSTER_META.get());
 
         Map<TopicPartition, List<ConsumerRecord<String, String>>> recordMap = new HashMap<>();
+        Map<TopicPartition, OffsetAndMetadata> nextOffsets = new HashMap<>();
+
         for (TopicPartition tp : records.partitions()) {
             List<ConsumerRecord<String, String>> lst = new ArrayList<>();
+            long nextOffset = 0;
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/test/MockConsumerInterceptor.java:80]: Have you considered the case here where the records for the topic partition is empty? Won't you return "next offset 0" in this case?

**Review by lucasbru — COMMENTED** [review]
Hi Alieh, thanks for the PR!

I left some comments.

Also, you committed a file called [clients/.attach_pid8858](https://github.com/apache/kafka/pull/17414/files#diff-27d1a6038c30c0a1ddc84b0ec60c461444bff84142060a8bc2d92b444f9fea09)

that you should remove from the PR

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 15]
```diff
@@ -21,23 +21,33 @@
...
-    public static final ConsumerRecords<Object, Object> EMPTY = new ConsumerRecords<>(Collections.emptyMap());
+    public static final ConsumerRecords<Object, Object> EMPTY = new ConsumerRecords<>(Collections.emptyMap(), Collections.emptyMap());
 
     private final Map<TopicPartition, List<ConsumerRecord<K, V>>> records;
+    private final Map<TopicPartition, OffsetAndMetadata> nextOffsets;
 
+    @Deprecated
     public ConsumerRecords(Map<TopicPartition, List<ConsumerRecord<K, V>>> records) {
         this.records = records;
+        this.nextOffsets = new HashMap<>();
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Since the `nextOffsets` map isn't modified after use, could this be:

```suggestion
        this.nextOffsets = Collections.emptyMap();
```

Or, the entire constructor could be:

```java
        this(nextOffsets, Collections.emptyMap());
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 16]
```diff
@@ -21,23 +21,33 @@
...
+    }
+
+    public ConsumerRecords(Map<TopicPartition, List<ConsumerRecord<K, V>>> records, final Map<TopicPartition, OffsetAndMetadata> nextOffsets) {
+        this.records = records;
+        this.nextOffsets = nextOffsets;
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Should this be immutable?

```suggestion
        this.nextOffsets = Collections.unmodifiableMap(nextOffsets);
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java (hunk 1 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: In her defense, the comments for the existing `records()` method are similarly terse 🤷‍♂️

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java [hunk 17]
```diff
@@ -229,6 +229,8 @@ public synchronized ConsumerRecords<K, V> poll(final Duration timeout) {
 
         // update the consumed offset
         final Map<TopicPartition, List<ConsumerRecord<K, V>>> results = new HashMap<>();
+        final Map<TopicPartition, OffsetAndMetadata> nextOffsetAndMetadata = new HashMap<>();
+
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: Super nit: remove the extra newline:

```suggestion
        final Map<TopicPartition, OffsetAndMetadata> nextOffsetAndMetadata = new HashMap<>();
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java (hunk 12 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java]: Any reason to not leave this line as-is so that it the single-arg `ConsumerRecords` is used?

**Review by kirktrue — CHANGES_REQUESTED** [review]
Thanks for the PR, @aliehsaeedii!

I didn't get as far as the unit tests, but most of my feedback are little nit-picks or echoing what @lucasbru already caught. I'll try to make a pass at the unit tests.

Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java (hunk 12 above)
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java]: >Any reason to not leave this line as-is so that it the single-arg ConsumerRecords is used?

since the current constructor with single-arg is marked as `@Deprecated`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java (hunk 3 above)
Comment (aliehsaeedii) [reply clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java]: to resolve the conflict with `trunk`. Maybe rebasing was a better approach.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java (hunk 12 above)
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java]: > Trivial, but please test this in `ShareConsumerImplTest`.

Mmm. `ShareConsumerImplTest` does not test the returned records even.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerInterceptorsTest.java (hunk 13 above)
Comment (aliehsaeedii) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerInterceptorsTest.java:145]: >those topic partitions are filtered out?

but we still keep the filtered out tps in `nextOffsets` map.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 18]
```diff
@@ -32,12 +32,19 @@
...
+    private final Map<TopicPartition, OffsetAndMetadata> nextOffsets;
 
+    @Deprecated
     public ConsumerRecords(Map<TopicPartition, List<ConsumerRecord<K, V>>> records) {
+        this(records, Collections.emptyMap());
+    }
+
+    public ConsumerRecords(Map<TopicPartition, List<ConsumerRecord<K, V>>> records, final Map<TopicPartition, OffsetAndMetadata> nextOffsets) {
         this.records = records;
+        this.nextOffsets = Collections.unmodifiableMap(nextOffsets);
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Now, you are doing `Collections.unmodifiableMap` twice. Once in the getter, once in the contructor. I would remove it here.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java (hunk 11 above)
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:199]: Thanks for the update. 

One more thing that I noticed - not 100% sure if it matters. In this code, even if the `position` does not advance (i.e., we haven't really fetched anything), we will still return the corresponding "nextOffset". I suppose (without checking further) that this would mean that `partRecords` is empty here.

I'm not sure if we should include a TP that did not advance its position at all in `nextOffsets`. After all, we promise to return only those topic partitions that have fetched something?

This is also why I pressed on a more precise comment above -- for which topic partitions do we precisely include a next offset, and for which do we not include one?

Thoughts? @kirktrue @aliehsaeedii

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java [hunk 19]
```diff
@@ -95,6 +102,13 @@ public int numRecords() {
         return numRecords;
     }
 
+    /**
+     * @return the next offsets and metadata (last epochs is included)
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java]: nit: `last epoch` not `last epochs`

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 20]
```diff
@@ -53,6 +60,14 @@ public List<ConsumerRecord<K, V>> records(TopicPartition partition) {
             return Collections.unmodifiableList(recs);
     }
 
+    /**
+     * Get the next offsets and metadata corresponding to all partitions fetched from the last batch.
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Why `fetched from the last batch`? When I follow the code, it seems the result can include the offsets from multiple batches?

@kirktrue made a fair point that other comments are also not good in this class. But for this method specifically, I think it should be clear for which topic-partitions we include offsets and for which topic-partitions we don't, as this will define what offsets we commit, and committing tends to be important in Kafka. So it should be very clear what is the set of offsets returned by this method.

I think we can say that we include offsets for all topic-partitions for which the position advanced in this poll call. But then, we need to change the code a bit (below). 

For me, this is not a minor point.

**Review by lucasbru — COMMENTED** [review]
Thanks for the updates, Alieh! Mostly LGTM, a few more comments. Let's also hear from Kirk if he has any more comments.

**Review by lucasbru — COMMENTED** [review]
Thanks for the updates, Alieh! Mostly LGTM, a few more comments. Let's also hear from Kirk if he has any more comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java (hunk 11 above)
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:199]: >we promise to return only those topic partitions that have fetched something?

What is `something`? Can we distinguish a `real nothing` and a `control record`?

**Comment by aliehsaeedii** [comment]
@lucasbru 
>I suppose (without checking further) that this would mean that partRecords is empty here.

I assume when control records are returned, `partRecords` is empty while the position is advanced.

**Comment by lucasbru** [comment]
> I assume when control records are returned, partRecords is empty while the position is advanced.

You are right. Anyway, I'd assume we add a TP to `nextOffsets` if the position advanced.

I can have a look again tomorrow or next week to see if all comments have been addressed, and potentially merge it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java (hunk 18 above)
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java]: Now that we have Java 11, this can be `Map.copyOf` instead of `Collections.unmodifiableMap`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java [hunk 21]
```diff
@@ -95,6 +102,13 @@ public int numRecords() {
...
+     */
+    public Map<TopicPartition, OffsetAndMetadata> nextOffsetAndMetadata() {
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java]: I'd probably just call this method `nextOffsets` to match what you have in `ConsumerRecords.nextOffsets`. You can also copy the description from that method with identical words. Seems like less effort.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java [hunk 22]
```diff
@@ -95,6 +102,13 @@ public int numRecords() {
...
+        return Collections.unmodifiableMap(nextOffsetAndMetadata);
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java]: `Map.copyOf`

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java [hunk 23]
```diff
@@ -1821,15 +1826,19 @@ private Answer<ConsumerRecords<byte[], byte[]>> expectConsumerPoll(final int num
...
+            }
             recordsReturnedTp1 += numMessages;
-            return new ConsumerRecords<>(
-                    numMessages > 0 ?
-                            Collections.singletonMap(new TopicPartition(TOPIC, PARTITION), records)
-                            : Collections.emptyMap()
-            );
+            final TopicPartition tp = new TopicPartition(TOPIC, PARTITION);
+            if (numMessages > 0) {
+                return new ConsumerRecords<>(Collections.singletonMap(tp, records),
```
Comment (AndrewJSchofield) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java]: This can be `Map.of(tp, records)`.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ErrorHandlingTaskTest.java [hunk 24]
```diff
@@ -482,8 +483,8 @@ private void createSourceTask(TargetState initialState, RetryWithToleranceOperat
     }
 
     private ConsumerRecords<byte[], byte[]> records(ConsumerRecord<byte[], byte[]> record) {
-        return new ConsumerRecords<>(Collections.singletonMap(
-                new TopicPartition(record.topic(), record.partition()), singletonList(record)));
+        final TopicPartition tp = new TopicPartition(record.topic(), record.partition());
+        return new ConsumerRecords<>(Collections.singletonMap(tp, singletonList(record)), Collections.singletonMap(tp, new OffsetAndMetadata(record.offset() + 1, record.leaderEpoch(), "")));
```
Comment (AndrewJSchofield) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ErrorHandlingTaskTest.java]: `new ConsumerRecords<L>(Map.of(tp, List.of(record)), Map.of(tp, new OffsetAndMetadata.....` is more concise in Java 11.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java (hunk 11 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:199]: Yes, but at the next level down. `CompletedFetch.nextFetchedRecord(FetchConfig)` is where we distinguish between normal records and control records. I think you'll probably need to be working inside `CompletedFetch`. This area of code is a bit nasty unfortunately. `nextFetchedRecord` is the key method to understand I think.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
This is trickier than it looks. I wonder whether keeping track of the information in the `FetchBuffer` makes most sense. This contains a queue of `CompletedFetch`, at most one per topic-partition. `CompletedFetch` is the class that actually processes the records and knows all about the offsets, and in particular control records.

**Comment by aliehsaeedii** [comment]
Thanks @AndrewJSchofield for the pointer. Yes, I followed the code, and as I stated before, when a control record is returned, the size of `partRecords` is `0` while the offset is updated.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java (hunk 22 above)
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java]: `Map.of` itself returns an immutable map. Maybe no need to create a copy?

**Comment by aliehsaeedii** [comment]
@lucasbru I think the PR is ready for the last round of review. Thanks in advance.

**Review by AndrewJSchofield — COMMENTED** [review]
I'd like specifically to see a test that, which I expect is a `FetchCollectorTest` that uses read-committed isolation level and lines up an initial `CompletedFetch` that has a dangling open transaction, and then a subsequent `CompletedFetch` that has just an abort marker, and than some real data, and checks that the records returned and offset metadata is what's expected in all cases. I expect you'll get an empty `Fetch` with updated offset metadata which advances as the consumer learns about the transaction's disposition.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 25]
```diff
@@ -620,10 +645,49 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
...
 
         assertTrue(fetch.isEmpty());
+        assertEquals(0, fetch.nextOffsets().size());
         verify(subscriptions).requestOffsetResetIfPartitionAssigned(topicPartition0);
         verify(fetchBuffer).setNextInLineFetch(null);
     }
 
+    @Test
+    public void testReadCommittedWithAbortedTransaction() {
+        buildDependencies();
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: Where are you configuring the fetch collector to use read_committed mode? I think you somehow have to extend `buildDependencies` to actually do what you state in the test name. I debugged your unit test, and it indeed runs in read_uncommmitted mode. So I don't think it's testing what Andrew was asking for. Check out how `fetchConfig` is constructed.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 26]
```diff
@@ -652,7 +653,7 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
 
     @Test
     public void testReadCommittedWithAbortedTransaction() {
-        buildDependencies();
+        buildDependencies(IsolationLevel.READ_UNCOMMITTED);
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: Don't you mean READ_COMMITTED here?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 27]
```diff
@@ -620,10 +646,49 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
...
 
+    @Test
+    public void testReadCommittedWithAbortedTransaction() {
+        buildDependencies(IsolationLevel.READ_COMMITTED);
+        int recordCount = 20;
+        assignAndSeek(topicAPartition0);
+
+
+        // Initial CompletedFetch
+        CompletedFetch initialCompletedFetch = completedFetchBuilder
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: I think you may want to also make the first completedFetch a transactional completed fetch. See `CompletedFetchTest` for an example. We should define the aborted transaction here.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 28]
```diff
@@ -620,10 +646,49 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
...
+            .recordCount(recordCount)
+            .build();
+        fetchBuffer.add(initialCompletedFetch);
+        Fetch<String, String> fetch = fetchCollector.collectFetch(fetchBuffer);
+
+        assertFalse(fetch.isEmpty());
+        assertEquals(recordCount, fetch.numRecords());
+        assertEquals(1, fetch.nextOffsets().size());
+        assertEquals(new OffsetAndMetadata(fetch.numRecords(), Optional.empty(), ""), fetch.nextOffsets().get(topicAPartition0));
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: nit: `fetch` is the result we are checking, don't use it on the left side of the assert, that's a bit confusing. Use `recordCount`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 29]
```diff
@@ -620,10 +646,49 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
...
         verify(fetchBuffer).setNextInLineFetch(null);
     }
 
+    @Test
+    public void testReadCommittedWithAbortedTransaction() {
+        buildDependencies(IsolationLevel.READ_COMMITTED);
+        int recordCount = 20;
+        assignAndSeek(topicAPartition0);
+
+
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: nit: double new line (all other tests use single new line)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 30]
```diff
@@ -620,10 +646,49 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
...
+            .build();
+        fetchBuffer.add(initialCompletedFetch);
+        Fetch<String, String> fetch = fetchCollector.collectFetch(fetchBuffer);
+
+        assertFalse(fetch.isEmpty());
+        assertEquals(recordCount, fetch.numRecords());
+        assertEquals(1, fetch.nextOffsets().size());
+        assertEquals(new OffsetAndMetadata(fetch.numRecords(), Optional.empty(), ""), fetch.nextOffsets().get(topicAPartition0));
+
+
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java:680]: nit: double new line (all other tests use single new line)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 31]
```diff
@@ -620,10 +646,49 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
...
+        // Subsequent CompletedFetch with abort marker
+        Records rawRecords = createTranscactionalRecords(ControlRecordType.ABORT, recordCount);
+        FetchResponseData.PartitionData partitionData = new FetchResponseData.PartitionData()
+            .setRecords(rawRecords);
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: If I understand correctly, the Andrew was asking for the ABORT control record to come first, before the other records.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 32]
```diff
@@ -840,7 +931,22 @@ private Records createRecords(final int recordCount) {
             for (int i = 0; i < recordCount; i++)
                 builder.append(0L, "key".getBytes(), ("value-" + i).getBytes());
 
-            return builder.build();
+            return builder;
+        }
+    }
+
+    private Records createTranscactionalRecords(ControlRecordType controlRecordType, int recordCount) {
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: nit: `Transactional` not `Trasncactional`

**Review by lucasbru — COMMENTED** [review]
Thanks for the updates, @aliehsaeedii ! I made another pass.

Editing my comment because I just understood how it works, ignore the previous version ;).

I think to implement the unit test that @AndrewJSchofield is asking for, we have to make the first `CompletedFetch` a transactional one, and we have to include an "abortedTransaction" field in the `RecordData` of the first `CompletedFetch`. It will make sure that all records up to the `ABORT` marker of the second `CompletedFetch` will be skipped.

Here is the document that describes the handling of fetches in more detail, in the "Consumer" section: https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit?tab=t.0

The expectation, I think, would be:

CompletedFetch1 contains 20 records and defines an aborted transaction
 --> nextOffset will be advanced, no record will be returned.

CompletedFetch2 contains a control record ABORT and then 20 records
 --> nextOffset will be advanced, 20 records will be returned.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java [hunk 33]
```diff
@@ -620,10 +650,61 @@ public void testCollectFetchInitializationOffsetOutOfRangeErrorWithOffsetReset()
...
+        fetch = fetchCollector.collectFetch(fetchBuffer);
+
+        // The Fetch object contains both data records and the advanced offset
+        assertFalse(fetch.isEmpty());
+        assertEquals(recordCount, fetch.numRecords());
+        assertEquals(1, fetch.nextOffsets().size());
+        assertEquals(new OffsetAndMetadata(startOffset + recordCount + 1, Optional.of(0), ""), fetch.nextOffsets().get(topicAPartition0));
+    }
+
+    private List<FetchResponseData.AbortedTransaction> createAbortedTransactions() {
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchCollectorTest.java]: I think `createAbortedTransaction` (singular) would be better. Also, I think you need to supply the `firstOffset` as a parameter which should be the offset of the first record of the transaction which is being aborted.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the read-committed test. Apart from a couple of nits on the test, lgtm.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm

**Review by lucasbru — APPROVED** [review]
LGTM, thanks!
